File: python/advanced/union_find.py
"""
Disjoint Set Union (DSU / Union-Find)
=======================================
A data structure that maintains a collection of disjoint (non-overlapping)
sets and supports two efficient operations:
    - find(x):     Determine which set *x* belongs to (returns representative).
    - union(x, y): Merge the sets containing *x* and *y*.

Two key optimizations:
    1. Path Compression (in find):  Flatten the tree so every node points
       directly to the root after a find call.
    2. Union by Rank (in union):    Always attach the smaller tree under the
       root of the taller tree to keep trees shallow.

With both optimizations, operations run in nearly O(1) amortized time:
    - find / union: O(α(n))  where α is the inverse-Ackermann function,
                              which is <= 4 for all practical n.
    - Space:        O(n)

Classic use cases:
    - Kruskal's Minimum Spanning Tree algorithm.
    - Detecting cycles in undirected graphs.
    - Connected components.
    - Network connectivity queries.
"""

from __future__ import annotations
from typing import Any, Dict, List, Optional, Tuple


class UnionFind:
    """
    Disjoint Set Union with path compression and union by rank.

    Supports any hashable element type (int, str, etc.).
    """

    def __init__(self) -> None:
        # parent[x] = representative of x's set (initially x itself).
        self._parent: Dict[Any, Any] = {}
        # rank[x] = upper bound on the height of x's subtree.
        self._rank: Dict[Any, int] = {}
        # Number of distinct sets.
        self._num_components: int = 0

    # ------------------------------------------------------------------
    # Make set
    # ------------------------------------------------------------------

    def add(self, x: Any) -> None:
        """
        Explicitly add element *x* to its own singleton set.
        No-op if *x* is already present.

        Time:  O(1)
        """
        if x not in self._parent:
            self._parent[x] = x
            self._rank[x] = 0
            self._num_components += 1

    def _ensure(self, x: Any) -> None:
        """Auto-add *x* if not present (lazy initialization)."""
        if x not in self._parent:
            self.add(x)

    # ------------------------------------------------------------------
    # Find (with path compression)
    # ------------------------------------------------------------------

    def find(self, x: Any) -> Any:
        """
        Return the representative (root) of the set containing *x*.

        Path Compression: on the way back up the recursion, every node on
        the path is made to point directly to the root.  This flattens the
        tree so future finds are O(1).

        Time:  O(α(n)) amortized
        """
        self._ensure(x)
        if self._parent[x] != x:
            # Recursive path compression: point x directly to root.
            self._parent[x] = self.find(self._parent[x])
        return self._parent[x]

    # ------------------------------------------------------------------
    # Union (by rank)
    # ------------------------------------------------------------------

    def union(self, x: Any, y: Any) -> bool:
        """
        Merge the sets containing *x* and *y*.
        Returns True if they were in different sets (merge happened),
        False if they were already in the same set.

        Union by Rank: attach the root with lower rank under the root with
        higher rank.  When ranks are equal, arbitrarily choose one root and
        increment its rank.  This keeps tree heights O(log n) without
        path compression, and nearly O(1) with it.

        Time:  O(α(n)) amortized
        """
        rx, ry = self.find(x), self.find(y)
        if rx == ry:
            return False  # already connected

        # Attach smaller rank tree under larger rank tree.
        if self._rank[rx] < self._rank[ry]:
            rx, ry = ry, rx  # ensure rx has the higher (or equal) rank
        self._parent[ry] = rx
        if self._rank[rx] == self._rank[ry]:
            self._rank[rx] += 1

        self._num_components -= 1
        return True

    # ------------------------------------------------------------------
    # Queries
    # ------------------------------------------------------------------

    def connected(self, x: Any, y: Any) -> bool:
        """
        Return True if *x* and *y* belong to the same set.

        Time:  O(α(n)) amortized
        """
        return self.find(x) == self.find(y)

    @property
    def num_components(self) -> int:
        """Number of distinct sets currently in the structure."""
        return self._num_components

    def component_of(self, x: Any) -> List[Any]:
        """
        Return all elements in the same set as *x*.

        Time:  O(n)  — iterates all elements.
        """
        root = self.find(x)
        return [e for e in self._parent if self.find(e) == root]

    def all_components(self) -> List[List[Any]]:
        """
        Return a list of lists, each being one connected component.

        Time:  O(n α(n))
        """
        groups: Dict[Any, List[Any]] = {}
        for e in self._parent:
            root = self.find(e)
            groups.setdefault(root, []).append(e)
        return list(groups.values())

    def __repr__(self) -> str:
        comps = self.all_components()
        return f"UnionFind(components={comps})"


# ----------------------------------------------------------------------
# Application: Kruskal's Minimum Spanning Tree
# ----------------------------------------------------------------------

def kruskal_mst(
    vertices: List[Any],
    edges: List[Tuple[int, Any, Any]],
) -> Tuple[List[Tuple[int, Any, Any]], int]:
    """
    Kruskal's algorithm for Minimum Spanning Tree.

    Parameters
    ----------
    vertices : list
        All vertex labels.
    edges : list of (weight, u, v)
        All edges as (weight, u, v) tuples.

    Returns
    -------
    mst_edges : list of (weight, u, v) in the MST
    total_weight : int

    Algorithm:
        1. Sort all edges by weight (greedy).
        2. For each edge (u, v), if u and v are in different components,
           add the edge to MST and union them.
        3. Stop when MST has V-1 edges.

    Time:  O(E log E) dominated by sorting.
    Space: O(V)
    """
    uf = UnionFind()
    for v in vertices:
        uf.add(v)

    mst_edges = []
    total_weight = 0

    for weight, u, v in sorted(edges):  # sort by weight
        if uf.union(u, v):              # no cycle
            mst_edges.append((weight, u, v))
            total_weight += weight
            if len(mst_edges) == len(vertices) - 1:
                break  # MST complete

    return mst_edges, total_weight


# ----------------------------------------------------------------------
# Demo
# ----------------------------------------------------------------------

if __name__ == "__main__":
    print("=== Union-Find ===")
    uf = UnionFind()
    for x in range(1, 8):
        uf.add(x)

    unions = [(1, 2), (2, 3), (4, 5), (6, 7)]
    for a, b in unions:
        uf.union(a, b)
        print(f"union({a}, {b})")

    print()
    print("connected(1, 3):", uf.connected(1, 3))  # True
    print("connected(1, 4):", uf.connected(1, 4))  # False
    print("num_components :", uf.num_components)   # 4
    print("all_components :", sorted(sorted(c) for c in uf.all_components()))

    uf.union(3, 4)
    print("\nAfter union(3, 4):")
    print("connected(1, 5):", uf.connected(1, 5))  # True
    print("num_components :", uf.num_components)   # 3

    print()
    print("=== Kruskal's MST ===")
    vertices = ["A", "B", "C", "D", "E"]
    edges = [
        (1, "A", "B"),
        (3, "A", "C"),
        (4, "B", "C"),
        (2, "B", "D"),
        (5, "C", "E"),
        (7, "D", "E"),
        (6, "B", "E"),
    ]
    mst, cost = kruskal_mst(vertices, edges)
    print("MST edges (weight, u, v):")
    for w, u, v in mst:
        print(f"  {u} — {v}  (weight={w})")
    print("Total MST cost:", cost)

File: java/advanced/UnionFind.java
package advanced;

import java.util.Arrays;

/**
 * Disjoint Set Union (Union-Find) with path compression and union by rank.
 *
 * <p>Supports {@code n} elements identified by integers in [0, n-1].
 *
 * <p>Both optimisations together yield an inverse-Ackermann amortised cost:
 * <ul>
 *   <li>find  – O(α(n)) amortised ≈ O(1) practically</li>
 *   <li>union – O(α(n)) amortised ≈ O(1) practically</li>
 * </ul>
 *
 * <p>Space complexity: O(n).
 */
public class UnionFind {

    // -------------------------------------------------------------------------
    // Fields
    // -------------------------------------------------------------------------

    private final int[] parent; // parent[i] = parent of element i; parent[root] = root
    private final int[] rank;   // rank[i] = upper bound on the height of the subtree rooted at i
    private int numComponents;  // number of disjoint components

    // -------------------------------------------------------------------------
    // Constructor
    // -------------------------------------------------------------------------

    /**
     * Creates a Union-Find structure for {@code n} elements.
     * Initially every element is its own component (n components).
     *
     * <p>Time: O(n) | Space: O(n).
     *
     * @param n number of elements (n >= 1)
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public UnionFind(int n) {
        if (n < 1) throw new IllegalArgumentException("n must be >= 1");
        parent = new int[n];
        rank   = new int[n];
        numComponents = n;
        for (int i = 0; i < n; i++) parent[i] = i; // each element is its own root
    }

    // -------------------------------------------------------------------------
    // find (with path compression)
    // -------------------------------------------------------------------------

    /**
     * Returns the representative (root) of the component containing element {@code x}.
     *
     * <p><em>Path compression</em>: every node on the path from {@code x} to the root
     * is directly attached to the root, flattening the tree for future calls.
     *
     * <p>Time: O(α(n)) amortised | Space: O(1) iterative.
     *
     * @param x element index (0-indexed)
     * @return root representative of x's component
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public int find(int x) {
        checkElement(x);
        // Path compression – iterative two-pass variant
        int root = x;
        while (parent[root] != root) root = parent[root];
        // Second pass: point everything on the path directly to root
        while (parent[x] != root) {
            int next = parent[x];
            parent[x] = root;
            x = next;
        }
        return root;
    }

    // -------------------------------------------------------------------------
    // union (by rank)
    // -------------------------------------------------------------------------

    /**
     * Merges the components containing {@code x} and {@code y}.
     *
     * <p><em>Union by rank</em>: the root of the smaller-rank tree is attached under
     * the root of the larger-rank tree, keeping the trees shallow.
     *
     * <p>Time: O(α(n)) amortised | Space: O(1).
     *
     * @param x first element
     * @param y second element
     * @return {@code true} if the two elements were in different components (a merge occurred)
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public boolean union(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        if (rootX == rootY) return false; // already in same component

        // Attach smaller-rank tree under larger-rank tree
        if (rank[rootX] < rank[rootY]) {
            parent[rootX] = rootY;
        } else if (rank[rootX] > rank[rootY]) {
            parent[rootY] = rootX;
        } else {
            // Same rank: pick one as root, increment its rank
            parent[rootY] = rootX;
            rank[rootX]++;
        }
        numComponents--;
        return true;
    }

    // -------------------------------------------------------------------------
    // connected
    // -------------------------------------------------------------------------

    /**
     * Returns {@code true} if {@code x} and {@code y} belong to the same component.
     *
     * <p>Time: O(α(n)) amortised | Space: O(1).
     *
     * @param x first element
     * @param y second element
     * @return {@code true} if connected
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public boolean connected(int x, int y) {
        return find(x) == find(y);
    }

    // -------------------------------------------------------------------------
    // getNumComponents
    // -------------------------------------------------------------------------

    /**
     * Returns the current number of disjoint components.
     *
     * <p>Time: O(1) | Space: O(1).
     *
     * @return number of components
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public int getNumComponents() {
        return numComponents;
    }

    // -------------------------------------------------------------------------
    // size of a component
    // -------------------------------------------------------------------------

    /**
     * Returns the number of elements in the component containing {@code x}.
     *
     * <p>Time: O(n) | Space: O(1).  (For O(α(n)) maintain a size[] array instead.)
     *
     * @param x element index
     * @return size of x's component
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public int componentSize(int x) {
        int root = find(x);
        int count = 0;
        for (int i = 0; i < parent.length; i++) {
            if (find(i) == root) count++;
        }
        return count;
    }

    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    private void checkElement(int x) {
        if (x < 0 || x >= parent.length)
            throw new IndexOutOfBoundsException("Element " + x + " out of range [0, " + (parent.length - 1) + "]");
    }

    // -------------------------------------------------------------------------
    // toString
    // -------------------------------------------------------------------------

    /**
     * Returns a string showing the parent array, rank array, and component groups.
     *
     * <p>Time: O(n log n) | Space: O(n).
     *
     * @return multi-line Union-Find state string
     */
    @Override
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public String toString() {
        int n = parent.length;
        StringBuilder sb = new StringBuilder();
        sb.append("UnionFind (n=").append(n)
          .append(", components=").append(numComponents).append("):\n");

        // Compute current roots (after path compression may update parent)
        int[] roots = new int[n];
        for (int i = 0; i < n; i++) roots[i] = find(i);

        sb.append("  index  : ");
        for (int i = 0; i < n; i++) sb.append(String.format("%3d", i));
        sb.append("\n  parent : ");
        for (int i = 0; i < n; i++) sb.append(String.format("%3d", parent[i]));
        sb.append("\n  rank   : ");
        for (int i = 0; i < n; i++) sb.append(String.format("%3d", rank[i]));
        sb.append("\n  root   : ");
        for (int i = 0; i < n; i++) sb.append(String.format("%3d", roots[i]));
        sb.append("\n");

        // Group elements by component
        sb.append("  Components:\n");
        java.util.Map<Integer, java.util.List<Integer>> groups = new java.util.TreeMap<>();
        for (int i = 0; i < n; i++) {
            groups.computeIfAbsent(roots[i], k -> new java.util.ArrayList<>()).add(i);
        }
        for (java.util.Map.Entry<Integer, java.util.List<Integer>> e : groups.entrySet()) {
            sb.append("    root=").append(e.getKey()).append(" -> ").append(e.getValue()).append("\n");
        }
        return sb.toString();
    }

    // -------------------------------------------------------------------------
    // Main – demo
    // -------------------------------------------------------------------------

    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public static void main(String[] args) {
        System.out.println("=== Union-Find Demo ===\n");

        UnionFind uf = new UnionFind(10);
        System.out.println("Initial state (10 elements):");
        System.out.println(uf);

        // Union operations
        System.out.println("union(0,1), union(2,3), union(4,5)");
        uf.union(0, 1);
        uf.union(2, 3);
        uf.union(4, 5);
        System.out.println(uf);

        System.out.println("union(6,7), union(8,9)");
        uf.union(6, 7);
        uf.union(8, 9);
        System.out.println(uf);

        System.out.println("union(0,2), union(4,6)");
        uf.union(0, 2);
        uf.union(4, 6);
        System.out.println(uf);

        System.out.println("union(0,4) -- merges two big components");
        uf.union(0, 4);
        System.out.println(uf);

        // Queries
        System.out.println("connected(1, 5) : " + uf.connected(1, 5) + "  (expected true)");
        System.out.println("connected(1, 8) : " + uf.connected(1, 8) + "  (expected false)");
        System.out.println("find(3)         : " + uf.find(3));
        System.out.println("getNumComponents: " + uf.getNumComponents() + "  (expected 3)");
        System.out.println("componentSize(0): " + uf.componentSize(0) + "  (expected 8)");

        // Path compression visible
        System.out.println("\nAfter find calls (path compression flattens parent array):");
        System.out.println(uf);
    }
}
